Feature/st 50 library js structure

pgs-gmartowski opened the pull request.

pgs-gmartowski commented on .stylelintrc:
```diff
@@ -33,6 +33,7 @@
       }
     ],
     "unit-whitelist": [
+      "px",
```
info: I prefer using px + %, rest of the units for me are redundant. Maybe rest of them here should be deleted ?

pgs-kzasada replied:
We should avoid using px as they are static and in Responsive Web Design world everything relies on proportions. The best practice is to use relative units: rem and em. Moreover, if you set font-size in pixels and a user changes browser font-size, this change does not reflect. In my opinion, whitelist should look like this: `["rem", "em", "%", "s", "deg"]`.

pgs-kzasada replied:
@pgs-gmartowski, @james-simpson-ST, @aaron-moore-ST what do you think about the list I suggested?

pgs-gmartowski replied:
agreed 👍

pgs-gmartowski commented on src/core/imports/iframe.ts:
```diff
@@ -0,0 +1,20 @@
+const appEndpoint = 'http://localhost:8080';
```
info: all these localhost urls are only for example usage; we should consider any other solution for example html pages.

pgs-gmartowski commented on .travis.yml:
```diff
@@ -1,22 +1,22 @@
   include:
-    - stage: "Build"
-      name: "Compile for distribution"
+    - stage: 'Build'
```
info: I think there are some issues with Prettier or some of you guys have different settings, there is a single quote rule in configuration.

pgs-kzasada replied:
In Visual Studio Code it works with plugin https://github.com/prettier/prettier-vscode - you don't need to configure anything. In WebStorm you have to follow these instructions - https://www.jetbrains.com/help/webstorm/prettier.html.

pgs-gmartowski replied:
so, there is no need to add .prettierrc file into project ? I have literally 2 rules there.

pgs-kzasada replied:
> so, there is no need to add .prettierrc file into project ? I have literally 2 rules there.

This file is needed to change the default configuration.

pgs-gmartowski replied:
ok

pgs-gmartowski commented on components/creditCardNumber/src/index.ts:
```diff
@@ -0,0 +1,44 @@
+import './style.scss';
```
FYI - in components folder there are standalone components with separate webpack configurations - for now I don't see any other solution for simulating iframes in our library. Eventually this html files should be placed somewhere on ST servers. @pgs-kzasada what do you think about that ? Could we leave it as it is or we should think about sth else ?

pgs-kzasada replied:
In my opinion, we should avoid multiple configurations. I'm looking at how to configure Webpack differently.

pgs-gmartowski replied:
ok

pgs-kzasada replied:
@james-simpson-ST, in the end, we will have two files: st-client.js and st-main.js.
st-client.js - This is is the script which is added by the merchant to his page via <script> tag. It is responsible for initializing iframes.
st-main.js - This script is a part of our fields in iframes and contains all logic responsible for validation, communication with API, etc. I don't think it is needed to split this script to scripts each for one field.

pgs-gmartowski commented on components/creditCardNumber/package.json:
```diff
@@ -0,0 +1,32 @@
+{
```
info: configuration for credit card component (separated from core configuration just like the other two components)

pgs-kzasada replied:
Leave it as it is for now. I'll create another pull request with final version - with only one package.json and Webpack

pgs-gmartowski replied:
👍

pgs-gmartowski commented on webpack.common.js:
```diff
@@ -30,8 +31,21 @@ module.exports = {
+    new HtmlWebpackPlugin(),
     new HtmlWebpackPlugin({
-      title: 'JS Library',
+      filename: 'creditCard.html',
```
info: this is components configuration in chunks, but I can't add in this type of configuration .js or css files to html, so I add separate `components` folder.

pgs-kzasada commented on components/securityCode/src/style.scss:
```diff
@@ -0,0 +1,30 @@
+    border-bottom: 1px solid #32325d;
+    height: 35px;
+    color: #32325d;
+    padding-left: 25px;
```
Colours, heights, paddings should be stored as variables in a separate file - `_variables.scss`

pgs-gmartowski replied:
True. For now I just leave it as it is but, yes I agree.

pgs-kzasada commented on src/core/classes/Element.class.ts:
```diff
@@ -0,0 +1,84 @@
+  });
+};
+
+class Element {
```
Please remember to add TypeDoc annotations.

pgs-gmartowski replied:
ok, done in this file.

pgs-kzasada commented on src/core/classes/validation/ExpireDate.class.ts:
```diff
@@ -0,0 +1,38 @@
+    element.addEventListener('keyup', (event: any) => {
+      let length = element.value.length;
+      if (length < 5) {
+        if (event.keyCode < 47 || event.keyCode > 57) {
```
Assign numbers to variables with meaningful names. Avoid "magic numbers".

pgs-gmartowski replied:
true, done.

pgs-kzasada commented on src/core/helpers/expireDateListeners.ts:
```diff
@@ -0,0 +1,26 @@
+    let expirationDateInput = <HTMLInputElement>(
+      document.getElementById('expiration-date')
+    );
+    applyStylesToIframe('expiration-date', 'http://localhost:8082/?');
```
You should remove hardcoded localhost in the final implementation.

pgs-gmartowski replied:
yes, that looks very bad :( I move it to the separate file but we have to figure out how to serve those iframes in other way.

pgs-gmartowski replied:
done

james-simpson-ST commented on src/core/imports/cards.ts:
```diff
@@ -0,0 +1,56 @@
+  general: /(\\d{1,4})(\\d{1,4})?(\\d{1,4})?(\\d+)?/,
+};
+
+const creditCardSecurityCodeLength = {
```
Shouldn't this have a type declaration?

pgs-gmartowski replied:
Good point ! I didn't notice it.

james-simpson-ST commented on src/core/classes/validation/SecurityCode.class.ts:
```diff
@@ -0,0 +1,9 @@
+import Validation from './Validation.class';
+
+class SecurityCode extends Validation {
```
Should this contain the length validation and only allow numerical inputs? Or has this just not been done yet? I realise this is part of STJS-17 and I'm just not sure how much of that is in this review

pgs-gmartowski replied:
it's just the definition of class; I haven't done validation for security code yet.
I've finished implementation in this branch, next changes I'll make in another branch.

james-simpson-ST commented on src/core/imports/iframe.ts:
```diff
@@ -0,0 +1,41 @@
+  cardNumber: 'http://localhost:8081',
+  securityCode: 'http://localhost:8082',
+  expirationDate: 'http://localhost:8083',
+};
```
We have the field input iframes sorted however I haven't seen the "control" frame that we'll use for communicating with our webservices backend being created and added. Does this need adding somewhere into the framework?

pgs-gmartowski replied:
yes, it should be somewhere in library, I'll think about how to aggregate all these endpoints.

james-simpson-ST commented on test/Element.spec.ts:
```diff
@@ -0,0 +1,56 @@
+describe('Index placeholder test', () => {
+  //when
+  let element, mockObject: object, createNewElement: object;
+  const cardNumberUrl = 'http://localhost:8081',
```
Shouldn't these URLs come from iframe.ts so we're not duplicating these URLs?

pgs-gmartowski replied:
done

james-simpson-ST commented on src/index.ts:
```diff
@@ -1 +1,60 @@
-import './style.scss';
+import '../examples/example.scss';
```
Previously Lee mentioned having a lightweight layer which basically asynchronously pulls in the majority of this code so that merchants stay up to date with our library. Has this been accounted for yet?

pgs-gmartowski replied:
yes, that's only the example. As soon as we figure out how to fix these problems with Webpack and his four standalone instances, we will change the approach to that,

pgs-gmartowski commented on components/expirationDate/templates/index.html:
```diff
@@ -0,0 +1,34 @@
+<!DOCTYPE html>
```
for now I see two solutions: 
- using some template engine like Handlebars
- Vanilla JS components
The second one is IMO better option.

james-simpson-ST replied:
Yes I agree JS components feels better as long as we don't want to consider a framework but I don't feel for the amount of UX we'll have that this will be beneficial. Does this feel like it needs to be it's own subtask to implement or will this be something that is easily implemented. I've never used JS components before to know the level of work involved

pgs-gmartowski replied:
I will check this topic today, but I think separate subtask is a good idea.

pgs-gmartowski replied:
@james-simpson-ST Actually we already have such task in JIRA :) https://securetrading.atlassian.net/browse/STJS-16
I'll rename it into sth more understandable :)

pgs-gmartowski commented on src/core/classes/Element.class.ts:
```diff
@@ -0,0 +1,84 @@
+
+const { cardNumber, securityCode, expirationDate } = iframesEndpoints;
+
+const RegisterElements = (fields: HTMLElement[], form: string) => {
```
I used to have `70% coverage rule` in previous projects, but frankly I was always focused on testing all of the crucial methods, not on the percentage. 

Unit tests are for me always part of 'definition of ready', doesn't matter if I use TDD or not.

james-simpson-ST replied:
Okay so in order for this task to be completed and "ready" would you agree that we should have unit tests for any of the functions with logic in which case they need to be completed before the review can be approved

pgs-gmartowski replied:
I'll talk with Kamil how to split these branch into smaller ones - I think cover all functions in current status could take some time.

pgs-gmartowski replied:
But generally yes, I agree, PR shouldn't be accepted without UT with proper coverage :)

pgs-gmartowski commented:
@pgs-kzasada, @aaron-moore-ST , @james-simpson-ST 
I've removed all the changes which had nothing in common with lib structure and I moved them into separate branches:

-  **feature/ST-59-expiration-date**
-  **feature/ST-60-security-code**
-  **feature/ST-69-communication-iframe-view**

I will create pull requests todays afternoon.

pgs-kzasada commented on components/expirationDate/templates/index.html:
```diff
@@ -0,0 +1,34 @@
+        />
+      </div>
+      <p id="received-message"></p>
+      <button
```
You are right, James. This is a work-in-progress version.

pgs-kzasada commented on components/expirationDate/webpack.dev.js:
```diff
@@ -0,0 +1,12 @@
+  devtool: "inline-source-map",
+  devServer: {
+    contentBase: "./dist",
+    port: 8082,
```
It will be hosted on a single server. I'm changing it now.
